autoload/ale/handlers/rust.vim

```diff
@@ -50,7 +50,7 @@ function! ale#handlers#rust#HandleRustErrorsForFile(buffer, full_filename, lines
 
             if (
             \   l:span.is_primary
-            \   && (l:span_filename ==# l:filename || l:span_filename ==# '<anon>')
+            \   && (a:full_filename =~ (l:span.file_name . '$') || l:span_filename ==# '<anon>')
```

Comment:
Could you try using `ale#path#IsBufferPath(a:buffer, l:span.file_name)` here, and let me know if that works? Matching with regex won't work in every case, as some other files might be similar enough that they match.